napari/plugins/_builtins.py

```diff
@@ -29,3 +36,125 @@ def napari_get_reader(path: Union[str, List[str]]) -> ReaderFunction:
         function that returns layer_data to be handed to viewer._add_layer_data
     """
     return image_reader_to_layerdata_reader(magic_imread)
+
+
+@napari_hook_implementation(trylast=True)
+def napari_write_image(path: str, data: Any, meta: dict) -> bool:
+    """Our internal fallback image writer at the end of the plugin chain.
+
+    Parameters
+    ----------
+    path : str
+        Path to file, directory, or resource (like a URL).
+    data : array or list of array
+        Image data. Can be N dimensional. If the last dimension has length
+        3 or 4 can be interpreted as RGB or RGBA if rgb is `True`. If a
+        list and arrays are decreasing in shape then the data is from an image
+        pyramid.
+    meta : dict
+        Image metadata.
+
+    Returns
+    -------
+    bool : Return True if data is successfully written.
+    """
+    imsave(path, data)
+    return True
+
+
+@napari_hook_implementation(trylast=True)
+def napari_write_points(path: str, data: Any, meta: dict) -> bool:
+    """Our internal fallback points writer at the end of the plugin chain.
+
+    Append `.csv` extension to the filename if it is not already there.
+
+    Parameters
+    ----------
+    path : str
+        Path to file, directory, or resource (like a URL).
+    data : array (N, D)
+        Coordinates for N points in D dimensions.
+    meta : dict
+        Points metadata.
+
+    Returns
+    -------
+    bool : Return True if data is successfully written.
+    """
+    ext = os.path.splitext(path)[1]
+    if ext != '.csv':
+        path = path + '.csv'
+
+    # construct table from data
+    table = []
+    for row in data:
+        table.append(list(row))
```

Comment:
```suggestion
    table = list(map(list, data))
```

But why isn't our own (?) `write_csv` capable of doing this?

Reply:
... Indeed it is, so this whole thing can be removed?

Reply:
Finally, does this work with our properties layer? I think it should? Specifically:

- if the points layer has properties, they should be included in the table
- the coordinates data should include axis names as the column headers, now that we have them. It is fine if they are just integers as is done by default. I would also be ok with `coord-0`, `coord-1`, ..., or `coord-x`, `coord-y`, etc if they have provided names. Anyway that's a small decision, but the label names should be used.

Reply:
> the coordinates data should include axis names as the column headers, now that we have them. 

we don't have them - right now they reside on the viewer and not on the layer and so don't get passed in through the `meta` dict. This is true for both reading and writing.

@GenevieveBuckley took the approach of `dim-0` through `dim-N` in #941 which is where i took this from. I can copy over that approach here, it will make the transition when we have axis labels easier.

I can add the points `properties` too if the dict is not empty

Reply:
> > the coordinates data should include axis names as the column headers, now that we have them.
> 
> we don't have them - right now they reside on the viewer and not on the layer and so don't get passed in through the `meta` dict. This is true for both reading and writing.

This is the second time this has been mentioned, so I've made an issue here so it can get picked up later on (not now, but eventually): https://github.com/napari/napari/issues/1144